### PR TITLE
Construct side_effect_expr_function_callt in a non-deprecated way [blocks: #3800]

### PR DESCRIPTION
**Pull request by tautschnig**
The constructor requires four arguments.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: af8467af2a13632ec1b8de01973b60658dde29a8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98376413

**Review by kroening — APPROVED** [review]

File: src/goto-instrument/goto_program2code.cpp [hunk 1]
```diff
@@ -342,9 +342,11 @@ goto_programt::const_targett goto_program2codet::convert_assign_varargs(
       {this_va_list_expr});
     f.arguments().back().type().id(ID_gcc_builtin_va_list);
 
-    side_effect_expr_function_callt type_of;
-    type_of.function() =
-      symbol_exprt("__typeof__", code_typet({}, empty_typet()));
+    side_effect_expr_function_callt type_of(
```
Comment (smowton) [inline src/goto-instrument/goto_program2code.cpp:347]: How do we eventually get away with this apparently-wrong type? Comment here to say where the fixup takes place?

Comment (tautschnig) [reply src/goto-instrument/goto_program2code.cpp:347]: Comment added (fixup never happens, we simply don't care).

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 4243502373a5c4fb5763a0a4e67b1d475c7cdad0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98476947
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: f3f893312a3a63732ab7d8fa68d07d349da4ec4e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98607139
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7376880794722e7ca9821eab6208afcba94a50bd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98711807

File: src/cpp/cpp_constructor.cpp [hunk 2]
```diff
@@ -223,9 +223,9 @@ optionalt<codet> cpp_typecheckt::cpp_constructor(
 
     side_effect_expr_function_callt function_call(
       cpp_namet(constructor_name, source_location).as_expr(),
-      operands_tc);
-
-    function_call.add_source_location()=source_location;
+      operands_tc,
+      typet(),
```
Comment (martin-cs) [inline src/cpp/cpp_constructor.cpp]: Is the default constructor for typet() really supported / the best option?

**Review by martin-cs — APPROVED** [review]
My parts are fine.

File: src/cpp/cpp_constructor.cpp (hunk 2 above)
Comment (tautschnig) [reply src/cpp/cpp_constructor.cpp]: #3941 introduces `uninitialized_typet`, which might be a better option. I'll make this PR depend on #3941.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 7be76c487658eb9bee81c395afa02f502f50c6f5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98896055
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 28ea7a59235f53d94ce91b293c5fc09feb7190dc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99028690

**Pull request merged**